### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by a-s01**
- Adds basic validation to DTO classes
- Adds error handling
- Utilizes Actuator and SpringDoc OpenApi on Swagger base
- Enhances validation by custom annotations using ConstraintValidator
- MapStruct was utilized in the previous homework version

Additionally:
- Adds multilingual support for validation exceptions
- Adds Validation service for business rules accordance

### CODE REVIEW

File: src/main/java/com/epam/spring/library/controller/handler/ErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import javax.validation.ConstraintViolationException;
+import java.time.Instant;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.epam.spring.library.exception.ErrorType.*;
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/library/controller/handler/ErrorHandler.java]: Can be package-private

File: src/main/java/com/epam/spring/library/controller/handler/ErrorHandler.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public List<ErrorDTO> handleMethodArgumentNotValidException(
+            MethodArgumentNotValidException ex,
+            HandlerMethod hm) {
+        log(ex, hm);
+        return ex.getAllErrors()
+                 .stream()
+                 .map(e -> getErrorDTO(e.getDefaultMessage(), VALIDATION_ERROR))
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/library/controller/handler/ErrorHandler.java:38]: refactor `getErroDTO` to functional style - make it return a `Function`

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -1,6 +1,31 @@
...
 config:
   primary-language: "en"
   languages:
       - "ru"
       - "ua"
       - "en"
+  keepPeriod: 14
+  messageFile: "messages"
+
+spring:
```
Comment (KostyantynPanchenko) [inline src/main/resources/application.yml]: It is recommended specifying active profile as a program argument and not hardcode it in properties file
or did you mean
```java
spring:
  config:
    activate:
      on-profile: dev

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -1,6 +1,31 @@
...
+  profiles:
+    active: "dev"
+
+management:
+  server:
+    port: 9090 #required for actuator & swagger-ui working along
+  endpoints:
+    web:
+      exposure:
+        include: "health,info"
```
Comment (KostyantynPanchenko) [inline src/main/resources/application.yml]:
```java
management:
  endpoint:
    health:
      enabled: true
      show-details: always
    info:
      enabled: true
    loggers:
      enabled: true
  endpoints:
    enabled-by-default: false
    web:
      exposure:
        include:
          - loggers
          - health
          - info

**Review by KostyantynPanchenko — APPROVED** [review]
Well done

File: src/main/resources/application.yml (hunk 3 above)
Comment (a-s01) [reply src/main/resources/application.yml]: Thanks, I will use it as the app argument now

**Pull request merged**